tools/harness-automation/parse_topofile.py

```diff
@@ -0,0 +1,157 @@
+#    names of its contributors may be used to endorse or promote products
+#    derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+"""
+parse_topofile.py
+-----------------
+This script is used to parse TopologyConfig.txt file to list vendor device info
+when preparing for the Thread Certification testbed
+
+usage: parse_topofile.py [-h] [-f TOPO_FILE] [-c CASE_LIST [CASE_LIST ...]]
+
+    parse TopologyConfig file and list all devices by case
+
+    optional arguments:
+      -h, --help            show this help message and exit
+      -f TOPO_FILE          Topology config file (default: C:/GRL/Thread1.1/Thread
+                            _Harness/TestScripts/TopologyConfig.txt)
+      -c CASE_LIST [CASE_LIST ...]
+                            Test case list (e.g. 5.1.1 9.2.1, default: all)
+
+Examples:
+    1. Get case 5.1.1 vendor info
+       cmd:   python parse_topofile.py -f TopologyConfig.txt -c 5.1.1
+       output:
+       case 5.1.1:
+            role-vendor pair: [('Leader', 'ARM'), ('Router_1', 'ARM')]
+            needed vendor devices:{'ARM': 2}
+
+       Testbed needed vendor devices:{'ARM': 2}
+
+    2. Get case 5.1.1 and 5.2.1 vendor info
+       cmd:    python parse_topofile.py -f TopologyConfig.txt -c 5.1.1 5.2.1
+       output:
+       case 5.1.1:
+           role-vendor pair: [('Leader', 'ARM'), ('Router_2', 'ARM')]
+           needed vendor devices:{'ARM': 2}
+       case 5.2.1:
+           role-vendor pair: [('Leader', 'OpenThread'), ('REED_1', 'Kirale'), ('MED_1', 'SiLabs')]
+           needed vendor devices:{'OpenThread': 1, 'Kirale': 1, 'SiLabs': 1}
+
+       Testbed needed vendor devices:{'ARM': 2, 'OpenThread': 1, 'Kirale': 1, 'SiLabs': 1}
+
+"""
+
+import argparse
+import logging
+import re
+from collections import Counter
+
+logging.basicConfig(format='%(message)s', level=logging.DEBUG)
+MAX_VENDOR_DEVICE = 32
+
+
+def device_calculate(topo_file, case_list):
+    testbed_vendor_dict = Counter()
+    with open(topo_file, 'r') as f:
+        for line in f:
+
+            case_vendor_dict = Counter()
+
+            if not line:
+                break
+            line = line.strip()
+
+            try:
```

Comment:
add comment to give a complete example that one typical line would look like? thus making it easier for reviewers who may not be aware of TopologyConfig.txt?

Reply:
didn't see the complete example for a typical line